src/comm/ConnOpener.cc

```diff
@@ -399,26 +400,17 @@ Comm::ConnOpener::retrySleep()
 {
     Must(!calls_.sleep_);
     closeFd();
-    calls_.sleep_ = true;
-    eventAdd("Comm::ConnOpener::DelayedConnectRetry",
-             Comm::ConnOpener::DelayedConnectRetry,
-             new Pointer(this), 0.05, 0, false);
+    typedef NullaryMemFunT<Comm::ConnOpener> Dialer;
+    calls_.sleep_ = JobCallback(5, 4, Dialer, this, Comm::ConnOpener::restart);
+    EventScheduler::GetInstance()->schedule("Comm::ConnOpener::restart", calls_.sleep_, 0.05, 0);
 }
 
 /// cleans up this job sleep state
 void
 Comm::ConnOpener::cancelSleep()
 {
     if (calls_.sleep_) {
-        // It would be nice to delete the sleep event, but it might be out of
-        // the event queue and in the async queue already, so (a) we do not know
-        // whether we can safely delete the call ptr here and (b) eventDelete()
-        // will assert if the event went async. Thus, we let the event run so
-        // that it deletes the call ptr [after this job is gone]. Note that we
-        // are called only when the job ends so this "hanging event" will do
-        // nothing but deleting the call ptr.  TODO: Revise eventDelete() API.
-        // eventDelete(Comm::ConnOpener::DelayedConnectRetry, calls_.sleep);
-        calls_.sleep_ = false;
+        calls_.sleep_->cancel("Comm::ConnOpener::cancelSleep");
```

Comment:
Why only remove from the event queue and not the AsyncCall queue as well?

Reply:
> Why only remove from the event queue and not the AsyncCall queue as well?

I do not recommend removing calls from the AsyncCallQueue because the only way to guarantee that a given call object is inside AsyncCallQueue (and not some _other_ call queue) is a linear search through a potentially long AsyncCallQueue. AsyncCallQueue is invasive, and there are other similarly invasive queues that store call objects. It is too dangerous to simply assume that we will be removing the call object from the _right_ queue. IIRC, that assumption holds in the current PR code for the sleep call, but no invariants guarantee it will continue to hold as the code changes.

Fortunately, unlike the time-ordered long-term event queue this PR is dealing with, the AsyncCallQueue tolerates accumulation of cancelled calls very well: More calls do not make FIFO queuing more expensive, and queued cancelled calls disappear after a single main loop iteration.

If somebody wants to cleanup AsyncCallQueue debugging logs by handling cancelled calls better, they can do it without removing the call object from the queue (but that AsyncCallQueue debugging improvement should not be done in _this_ PR).

In summary, the correct removal is too costly and similar (minor) improvements can be achieved by other (cheap and more general) means.

Reply:
The event queue is much the same as the AsyncCall queue. Although less invasive they share all the performance related issues of linear search for explicit removal, easy skipping of canceled Call's at the head of queue.

The costs of leaving events is cycles to skip when adding new events with longer delays. IMO the best optimization for performance of that would be removal of canceled entries when they are encountered by the next add. That way the cancel does not have to walk all events up to their location, and future adds do not have to step over them.

Reply:
> The event queue is much the same as the AsyncCall queue. 

I have [detailed](https://github.com/squid-cache/squid/pull/892#discussion_r698123098) the _critical_ differences already: time-ordered and long-term versus FIFO and instantaneous.

> Although less invasive they share all the performance related issues of linear search for explicit removal

... except we do not need to remove anything from the AsyncCallQueue (as I have already [detailed](https://github.com/squid-cache/squid/pull/892#discussion_r698123098)).

BTW, the two queues are _equally_ invasive. That invasiveness may be the only design aspect they share, but we do not need to agree on that.

> The costs of leaving events is cycles to skip when adding new events with longer delays.

... and the cost of firing unwanted events (cbdata validity cannot be used to stop unwanted events from firing because cbdata may remain valid after the event becomes unwanted).

... and the memory and reporting/triage costs of keeping unwanted events.

> IMO the best optimization for performance of that would be removal of canceled entries when they are encountered by the next add. That way the cancel does not have to walk all events up to their location, and future adds do not have to step over them.

You are describing lazy garbage collection. There are grave problems with that optimization in Squid context. Fortunately, this PR is not about optimizing event queue performance, so we do not need to argue about any of this. My change request is about _preserving_ the existing event queue implementation strategy and usage pattern -- delete unwanted events. The official Squid code relies on that approach. This PR has no valid reasons for changing any of that.

If you want to spend time on optimizing the event queue that the current official code tries to keep very short, then we can discuss that idea elsewhere. Let's not enlarge this PR scope with that (unwanted IMO) change.

Reply:
> My change request is about _preserving_ the existing event queue implementation strategy and usage pattern -- delete unwanted events. The official Squid code relies on that approach. This PR has no valid reasons for changing any of that.

That is not true. This ConnOpener is one of several cases where the official code *does not* remove unwanted events. It left them in the event queue to be skipped when ScheduleCall() discovers the Job cbdata pointer to be invalid.

Reply:
> That is not true. This ConnOpener is one of several cases where the official code does not remove unwanted events. It left them in the event queue to be skipped when ScheduleCall() discovers the Job cbdata pointer to be invalid.

What I said is true. The ConnOpener event handling code has a well-documented bug that could not be fixed using the old/official API (as detailed in that bug documentation removed by this PR). The new API added in this PR makes that fix possible. This change request asks to implement that fix (in addition to removing bug documentation; that removal has happened already, bu the bug is still there). If you would rather not implement this fix, I will; just let me know.

Reply:
With simply the conversion to Call storage the calls_.sleep_ can be canceled at any time. Even if not in the event queue - which resolves the bug regardless of whether the event entry is removed or left in place.

For further evidence I have also converted the TcpLogger caller which is just as simple and did not remove events either. It is not documented as buggy. None of the following list (vast majority of event types) are ever removed from the event queue either.
```
cacheCleanupEventName
client_db
commHalfClosedCheck
commHandleWriteHelper
cpuProfiling
Digest
fqdncache_purgelru
idnsCheckQueue
ipcache_purgelru
IpcIoFile::CheckTimeouts
IpcIoFile::DiskerHandleMoreRequests
IpcIoFile::OpenTimeout
Ipc::UdsSender::DelayedRetry
mainReconfigureFinish
MaintainSwapSpace
memPoolCleanIdlePools
netdbExchangeStart
netdbSaveState
peerClearRR
peerCountMcastPeersDone
peerCountMcastPeersStart
peerDigestCheck
Rock::Rebuild
send_announce
SquidShutdown
SquidTerminate
statAvgTick
statObjects
storeCleanup
storeClientCopyEvent
storeDigestRebuildStart
storeDigestRebuildStep
storeDigestRewriteStart
storeDigestSwapOutStep
storeDirClean
storeLateRelease
storeMaintain
storeRebuild
UFS storeDirClean
wccpHereIam
```

Reply:
I do not understand why you are resisting this change request. I can dismantle the counter-arguments one-by-one, but I just do not see why we should be wasting time on arguing about this. The facts I [outlined](https://github.com/squid-cache/squid/pull/892#discussion_r698521029) are not going to disappear. The problem is real, well-known, and has caused bugs. The requested fix is straightforward, very beneficial, and can be viewed as in-scope. I can implement it if you prefer. Why argue?!

Reply:
> I do not understand why you are resisting this change request.

FTR: I have added the removal requested in 2603370. I just do not agree on the reason(s) provided for doing so being anywhere as relevant as you state. The legacy API function storage is fully capable of removal, but is not used that way. If we are going to formalize this requirement on all callers during the transition we need to be in agreement on why, so we get the implementation choices right.

Reply:
> If we are going to formalize this requirement on all callers during the transition we need to be in agreement on why

Cancelled events should be removed from the event queue because they can accumulate in the event queue (with time and/or under load) while the event queue cannot handle large number of queued events well:

1. event insertion uses linear search that may become too slow
2. event deletion uses linear search that may become too slow
3. cache manager event list reporting will contain too much noise (if it works at all)

Also, in general, when simple/safe/cheap/etc. garbage collection is possible at garbage creation time, it should be used at that time (as opposed to relying on delayed/lazy garbage collection). This rule reduces the total/maximum amount of waste in the system, penalizes the garbage creator (rather than some innocents), and avoids creation of costly delays/stalls (that delayed/lazy garbage collection is well known for). All these are sound engineering principles/goals.

Any objections to the above rules of thumb and their justification?

----

> The legacy API function storage is fully capable of removal, but is not used that way.

I am not sure what you mean by "capable" here, but it was not possible to use the legacy `event.{cc,h}` code to reliably cancel certain events (as [documented](https://github.com/squid-cache/squid/blob/master/src/comm/ConnOpener.cc#L413)). It was not an API problem or, to be more precise, it was not just the API problem. That is why at least some of the code that should be removing its stale events was not removing them.

Certain events were removable (and were removed as needed) using legacy code.